File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
    
class Tree:
    def __init__(self, root):
        self.root = root
    
    def preorder(self, node):
        print(node.data, end="")
        if (node.left != None):
            self.preorder(node.left)
        if (node.right != None):
            self.preorder(node.right)

    def inorder(self, node):
        if (node.left != None):
            self.inorder(node.left)
        print(node.data, end="")
        if (node.right != None):
            self.inorder(node.right)

    def postorder(self, node):
        if (node.left != None):
            self.postorder(node.left)
        if (node.right != None):
            self.postorder(node.right)
        print(node.data, end="")


if __name__ == "__main__":
    cnt = int(sys.stdin.readline())
    nodes = {}

    for i in range(cnt):
        name, left, right = sys.stdin.readline().split()

        if (name not in nodes):
            nodes[name] = Node(name)

        if (left != "." ):
            if (left not in nodes):
                nodes[left] = Node(left)
            nodes[name].left = nodes[left]
        
        if (right != "." ):
            if (right not in nodes):
                nodes[right] = Node(right)
            nodes[name].right = nodes[right]
    
    tree = Tree(nodes["A"])

    tree.preorder(tree.root)
    print()
    tree.inorder(tree.root)
    print()
    tree.postorder(tree.root)

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.*;
import java.util.*;

class Node {
    String name = "";
    String[] child;

    public Node(String name, String[] arr) {
        this.name = name;
        this.child = arr;
    }
}

public class Main {

    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static HashMap<String, Node> tree = new HashMap<>();
    public static boolean[] preorder;
    public static boolean[] inorder;
    public static boolean[] postorder;

    public static void main(String[] args) throws IOException {
        int n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            String node = st.nextToken();
            String left = st.nextToken();
            String right = st.nextToken();
            tree.put(node, new Node(node, new String[]{left, right}));
        }

        preorder = new boolean[n + 1];
        preorder("A");
        System.out.println();

        inorder = new boolean[n + 1];
        inorder("A");
        System.out.println();

        postorder = new boolean[n + 1];
        postorder("A");
    }

    // 루트 - 왼쪽 - 오른쪽
    private static void preorder(String nodeIdx) {
        Node node = tree.get(nodeIdx);
        System.out.print(node.name);
        for (String str : node.child) {
            if (str.equals(".")) {
                continue;
            }
            preorder(str);
        }
    }

    // 왼쪽 - 루트 - 오른쪽
    private static void inorder(String nodeIdx) {
        Node node = tree.get(nodeIdx);
        String left = node.child[0];
        String right = node.child[1];
        if (!left.equals(".")) {
            inorder(left);
        }
        System.out.print(nodeIdx);
        if (!right.equals(".")) {
            inorder(right);
        }
    }

    // 왼쪽 - 오른쪽 - 루트
    private static void postorder(String nodeIdx) {
        Node node = tree.get(nodeIdx);
        String left = node.child[0];
        String right = node.child[1];
        if (!left.equals(".")) {
            postorder(left);
        }
        if (!right.equals(".")) {
            postorder(right);
        }
        System.out.print(nodeIdx);
    }
}
